### PR TITLE
First shot at a unified proposal

### PR DESCRIPTION
**Pull request by huitema**
This derives from the discussions on issue #96

### CODE REVIEW

**Review by Yanmei-Liu — APPROVED** [review]
Look good to me.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:251]: This seems more related to the discussion in #101, not sure if it should be part of this PR.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:597]: I don't think per-path congestion control is specific to NULL CID case.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:602]: Actually, the beginning of this paragraph is not specific to the NULL CID case.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:606]:
```suggestion
which packet was sent to what path. The senders MUST use that information to
```

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:637]:
```suggestion
to a number space. When sending to NULL CID receivers, the same number space is
```

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:661]: why "mostly"?

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:595]:
```suggestion
### Congestion Control, Loss Recovery and NULL CID Considerations
```

**Review by qdeconinck — COMMENTED** [review]
I believe this PR is a good first proposal for such a unified solution. If such a solution can fit all the use cases, I'm happy to further work on this.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md:597]: Correct. Need to state that correctly. The point is, in the absence of CID, implementations must track the path over which the packet loss or the RTT measurement occurred, because the number space is shared.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md:602]: Yes.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md:661]: :-)
need to find another way of saying that without a number space per path this is harder to get right.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md:595]: Let's first apply @mirjak's suggestion to replace NULL by zero-length...

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md]: Removed "mostly"

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md]: Fixed.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md]: Fixed, per next comment.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md]: Reverted.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md]: Fixed

**Comment by huitema** [comment]
The commit [a230bc1](https://github.com/quicwg/multipath/pull/103/commits/a230bc17db0eabf78dffb6e0d02224ad851197da) applies the changes suggested in the previous reviews. With those changes, I think we have a good basis for discussing a unified solution in the working group.

**Review by Yanmei-Liu — APPROVED** [review]

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:137]:
```suggestion
sent to that node over different paths. All packets sent to a zero-length
```

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:142]:
```suggestion
use multiple paths when sending to a node using zero-length CID will
```

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:144]:
```suggestion
delivery on loss detection and congestion controller. If the receiver
```

**Review by qdeconinck — COMMENTED** [review]
Some suggestions to remove NULL CID mentions, but overall this seems a good basis for further discussion.

File: draft-ietf-quic-multipath.md
Comment (boucadair) [reply draft-ietf-quic-multipath.md:134]: RFC9000 says the following:

```
  An endpoint SHOULD NOT initiate migration with a peer that has
   requested a zero-length connection ID, because traffic over the new
   path might be trivially linkable to traffic over the old one.
```

I wonder if the above behavior has implications on maintaining multiple paths when  zero-length connection IDs are in used.

File: draft-ietf-quic-multipath.md
Comment (boucadair) [inline draft-ietf-quic-multipath.md:146]: Again, the "specific" thing is worth be exemplified.

File: draft-ietf-quic-multipath.md
Comment (boucadair) [inline draft-ietf-quic-multipath.md:145]:
```suggestion
reports ECN counters, they will contain the sum of markings received on all
```

**Review by boucadair — CHANGES_REQUESTED** [review]
Some very minor comments.

**Comment by huitema** [comment]
Thanks @boucadair for the review. I have applied the editorial changes. You are asking for something more regarding the "specific logic" required when supporting zero-length CID. There are in fact two pieces to that: logic at the receiver, i.e., the node that chose to receive packets with zero-length CID; and logic at the sender, i.e., the node that accepts to send data on multiple paths towards a node that uses zero-length CID. This is explained in details in the section "Using Zero-Length connection ID", so I guess what we need is a reference to that section.

File: draft-ietf-quic-multipath.md
Comment (huitema) [inline draft-ietf-quic-multipath.md:150]: This lines were added to address review by @boucadair

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md]: The new title is "Loss and Congestion Handling With Zero-Length CID". I hope it works.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md]: The RFC 9000 text suggests that only the node that is choosing to use zero length CID will initiate a new path -- typically a client. That seems fine. Third parties will be able to observe packets flowing between client and server as follow:

* path 1: between address 1 and port 1 on client, and server address and server port on server.
* path 2: between address 2 and port 2 on client, and server address and server port on server.

The addresses and ports will be visible whether the client uses null-length CID or not. We may want to say to point out that using different port numbers for each path may remove 16 bits of possible correlation data, but that applies to RFC 9000 just as well.

Comment (boucadair) [reply draft-ietf-quic-multipath.md]: I agree this is OK in typical scenarios. 

I leave it up to you to decide if it is worth to include a mention about this in the draft. Thanks, @huitema.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [inline draft-ietf-quic-multipath.md:140]:
```suggestion
the size of ACK frames. This proposal specifies algorithms for
```

**Review by qdeconinck — APPROVED** [review]
Small typo; for the remaining, LGTM.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:567]: I don't think the MUST and SHOULD here make a lot of sense, given there are more/other normative language later in this section. Also ice think your removed too much.... here is an alternative proposal for this section intro part:
```suggestion
If a zero-length connection ID is used, one packet number space
 for all paths. That means the packet sequence numbers are allocated from the common
 number space, so that, for example, packet number N could be sent
 on one path and packet number N+1 on another.
 
 In this case, ACK frames report the numbers of packets that have been received so far,
 regardless of the path on which they have been received. That means
 the senders needs to maintain an association between sent packet numbers
 and the path over which these packets were sent. This is necessary
 to implement per path congestion control, as explained
in {{zero-length-cid-loss-and-congestion}}.

Further, the receiver of packets with zero-length connection IDs should
handling of acknowledgements defined in {{sending-acknowledgements-and-handling-ranges}}.

ECN handing is discussed in {{ecn-handling}}, and
mitigation of the RTT measurement is further explained
in {{ack-delay-and-zero-length-cid-considerations}}. 

If a node
does not support this logic, it MUST limit its use of multipath
as explained in {{restricted-sending-to-zero-length-cid-peer}}.
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:631]:
```suggestion
### ECN and Zero-Length CID Considerations {#ecn-handing}
```

**Review by mirjak — CHANGES_REQUESTED** [review]
I made an editorial pass but am not done yet. @huitema maybe you can merge (or reject) already these changes and I finish my review tomorrow.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:637]:
```suggestion
ECN feedback in QUIC is provided based on counters in the ACK frame
(see {{Section  19.3.2. of QUIC-TRANSPORT}}). That means if an ACK
 frame acknowledges multiple packets, the ECN feedback cannot be accounted
 to a specific packet.

There are separate counters for each packet number space. However, sending to zero-length CID receivers, the same number space is used for multiple paths. Respectively, if an ACK frames acknowledges multiple
packets from different paths, the ECN feedback cannot unambiguously be assigned to
a path.
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:643]:
```suggestion
transmission on the sending path. If zero-length CID is used, 
the sending path is however ambiguous. Therefore, the sender MUST
treat a CE marking as a congestion signal on all sending paths.
```

Comment (mirjak) [reply draft-ietf-quic-multipath.md:643]: Note that this change is not only editorial as I added the clause that you only MUST decrease those paths where you actually got a packet acknowledged. This is slightly more complicated but you should have saved the information where which packet was sent anyway and this allows for optimisation in the ACK creation. Given this is a MUST, I think we need this clause.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:653]:
```suggestion
a peer using zero-length CID. The simplest posture is to only send
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:571]:
```suggestion

If zero-length CID and therefore also a single packet number space
is used by the sender, the receiver MAY send ACK frames instead
of MP-ACK frames to reduce overhead as the additional path ID field
will anyway always carry the same value.

```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:614]:
```suggestion
based on timers and on sequence numbers. When packets are sent over multiple paths,
loss detection must be adapted to allow for different RTTs on different paths.
When sending to zero-length CID receivers, packets sent on different paths may not be received in
order. Therefore senders cannot directly use the packet sequence numbers to
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:620]: I don't think I fully understand these last couple of sentences. Can you clarify?

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:668]:
```suggestion
## Using non-zero length CID and Multiple Packet Number Spaces

If packets contain a non-zero CID, each path has
```

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md]: This is outdated.

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md:683]: There is always the occasional probe, etc.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:635]:
```suggestion
on different paths may be received out of order. Therefore senders cannot
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:139]:
```suggestion
on multiple paths, out of order delivery is likely. This causes inflation of the number of
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:144]:
```suggestion
delivery on loss detection, congestion control, and ECN handling.
```

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:629]:
```suggestion
{{ecn-handling}} for issues on using ECN marks.)
```

File: draft-ietf-quic-multipath.md
Comment (huitema) [reply draft-ietf-quic-multipath.md:642]: Fixed.

File: draft-ietf-quic-multipath.md
Comment (mirjak) [inline draft-ietf-quic-multipath.md:138]:
```suggestion
zero-length CID by using the same packet number space on all paths.
```

**Review by yfmascgy — APPROVED** [review]

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:685]: That is what I understand from the current text, yes.

File: draft-ietf-quic-multipath.md
Comment (qdeconinck) [reply draft-ietf-quic-multipath.md:653]: Maybe should we create an issue referencing this and create a specific PR afterwards?

Comment (mirjak) [reply draft-ietf-quic-multipath.md:653]: Created issue #125

File: draft-ietf-quic-multipath.md
Comment (mirjak) [reply draft-ietf-quic-multipath.md:685]: Yes, that's the meaning. For me the text is clear. I will merge the PR now, however, if you think further clarification is needed please propose a separate PR.

**Pull request merged**
